src/osd/PGLog.cc

```diff
@@ -1277,6 +1277,7 @@ seastar::future<> PGLog::rebuild_missing_set_with_deletes_crimson(
 	  log_entry.version,
 	  eversion_t(),
 	  log_entry.is_delete());
+	return seastar::now();
```

Comment:
Returning `seastar::now()` will stop the iteration, why not returning `seastar::stop_iteration::no;` instead?
IIUC the iteration shouldn't stop on missing object.

Reply:
I thought error handlers that return void are treated as if they return seastar::now().  I'm also a bit confused -- I don't think this method should actually ever happen on a crimson cluster.  It's there to handle upgrading from a prior version of classic where we didn't record deleted objects in the missing set instead handling them inline during peering.  If we're actually triggering this in crimson, something's wrong.

Reply:
> I thought error handlers that return void are treated as if they return seastar::now(). 

Yeah, I thought so, too. But it looks errorator returns the original exception future if error handlers return void: https://github.com/ceph/ceph/blob/b3bbd9f53aa54dc1db3e8396b61cb0d55ccdd7b6/src/crimson/common/errorator.h#L288-L356

> I'm also a bit confused -- I don't think this method should actually ever happen on a crimson cluster. It's there to handle upgrading from a prior version of classic where we didn't record deleted objects in the missing set instead handling them inline during peering. If we're actually triggering this in crimson, something's wrong.

It seems that `PeeringListener::rebuild_missing_set_with_deletes` is called as long as the missing set `may_include_delete` and `CEPH_OSDMAP_RECOVERY_DELETES` is not set. Am I missing anything?

Reply:
Right, it's just that that combination of conditions should never happen for a crimson cluster -- it should only happen during upgrade from the prior version of the missing set.  Sounds like the change is still correct though since we did handle the exception.